File: cpp/144_BinaryTreePreorderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 
 //Recursive
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> ret;
        if(root == NULL) return ret;
        ret.push_back(root->val);
        vector<int> leftRet = preorderTraversal(root->left);
        ret.insert(ret.end(),leftRet.begin(),leftRet.end());
        vector<int> rightRet = preorderTraversal(root->right);
        ret.insert(ret.end(),rightRet.begin(),rightRet.end());
        return ret;
    }
};

//Iterative
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> ret;
        stack<TreeNode *> sTree;
        sTree.push(root);
        while(sTree.size()){
            auto t = sTree.top();
            sTree.pop();
            if(t != NULL){
                ret.push_back(t->val);
                if(t->right) sTree.push(t->right);
                if(t->left) sTree.push(t->left);
            }
        }
        return ret;
    }
};

File: js/144_BinaryTreePreorderTraversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
 
//Recursive
var preorderTraversal = function(root) {
    var ret = [];
    if(root === null) return ret;
    ret.push(root.val);
    var leftRet = preorderTraversal(root.left);
    ret = ret.concat(leftRet);
    leftRet = null;
    var rightRet = preorderTraversal(root.right);
    ret = ret.concat(rightRet);
    rightRet = null;
    
    return ret;
};

//Iterative
var preorderTraversal = function(root) {
    var ret = [];
    var stack = [];
    stack.push(root);
    while(stack.length > 0){
        var t = stack.pop();
        if(t !== null){
            ret.push(t.val);
            if(t.right) stack.push(t.right);
            if(t.left) stack.push(t.left);
        }
    }
    return ret;
};